Add public APIs to Access Underlying `cudf` and `pandas` Objects from `cudf.pandas` Proxy Objects (#17629)
rapidsai/cudf: docs/cudf/source/cudf_pandas/faq.md

blog](https://medium.com/rapids-ai/easy-cpu-gpu-arrays-and-dataframes-run-your-dask-code-where-youd-like-e349d92351d)) and the [RAPIDS Accelerator for Apache Spark](https://nvidia.github.io/spark-rapids/)
provides a similar configuration-based plugin for Spark.

## Recommendation for libraries that are type aware.

When working with `cudf.pandas` proxy objects, it is important to access the real underlying objects to ensure compatibility with libraries that are `cudf` or `pandas` aware. You can use the following methods to retrieve the actual `cudf` or `pandas` objects:

- `as_gpu_object()`: This method returns the fast `cudf` object from the proxy.
- `as_cpu_object()`: This method returns the slow `pandas` object from the proxy.

Here is an example of how to use these methods:

```python
# Assuming `proxy_obj` is a cudf.pandas proxy object
fast_obj = proxy_obj.as_gpu_object()
slow_obj = proxy_obj.as_cpu_object()

# Now you can use `fast_obj` and `slow_obj` with libraries that are cudf or pandas aware
```

> 

Author: yes!

blog](https://medium.com/rapids-ai/easy-cpu-gpu-arrays-and-dataframes-run-your-dask-code-where-youd-like-e349d92351d)) and the [RAPIDS Accelerator for Apache Spark](https://nvidia.github.io/spark-rapids/)
provides a similar configuration-based plugin for Spark.

## How do I know if an object is a `cudf.pandas` proxy object?

To determine if an object is a `cudf.pandas` proxy object, you can use the `is_cudf_pandas_obj` API. This function checks if the given object is a proxy object that wraps either a `cudf` or `pandas` object. Here is an example of how to use this API:

```python
from cudf.pandas import is_cudf_pandas_obj

obj = ...  # Your object here
if is_cudf_pandas_obj(obj):
    print("The object is a cudf.pandas proxy object.")
else:
    print("The object is not a cudf.pandas proxy object.")
```

> There are various APIs to detect `Series`, `DataFrame`, `Index` and `ndarray`'s separately:

Reviewer: ```suggestion
There are APIs to detect `Series`, `DataFrame`, `Index`, and `ndarray` objects separately:
```

blog](https://medium.com/rapids-ai/easy-cpu-gpu-arrays-and-dataframes-run-your-dask-code-where-youd-like-e349d92351d)) and the [RAPIDS Accelerator for Apache Spark](https://nvidia.github.io/spark-rapids/)
provides a similar configuration-based plugin for Spark.

## How do I know if an object is a `cudf.pandas` proxy object?

To determine if an object is a `cudf.pandas` proxy object, you can use the `is_cudf_pandas_obj` API. This function checks if the given object is a proxy object that wraps either a `cudf` or `pandas` object. Here is an example of how to use this API:

```python
from cudf.pandas import is_cudf_pandas_obj

obj = ...  # Your object here

> if is_cudf_pandas_obj(obj):

Reviewer 2: Question: do we think this would feel more familiar to users to mimic `isinstance` more closely? So something `cudf_pandas_isinstance(obj, typename)`? Depending on whether cudf.pandas is active we should know whether the type we're ingesting is going to be either a cudf.pandas type or a real cudf/pandas type. cudf.pandas has this information even if cudf internally does not. If we do this, then we don't need four separate functions below.

Reviewer: I think that is a nicer API.

Author: I like it aswell. Updated it.

> if isinstance_cudf_pandas(obj, "Series"):

Reviewer: Normally `isinstance` takes a class, not a string. Is that an option here, too?

Author: responded here; https://github.com/rapidsai/cudf/pull/17629#discussion_r1932435799

    print("The object is a cudf.pandas proxy object.")
else:
    print("The object is not a cudf.pandas proxy object.")
```

There are APIs to detect `Series`, `DataFrame`, `Index`, and `ndarray` objects separately:

* `is_cudf_pandas_series`: Detects if the object is a `cudf.pandas` proxy `Series`.
* `is_cudf_pandas_dataframe`: Detects if the object is a `cudf.pandas` proxy `DataFrame`.
* `is_cudf_pandas_index`: Detects if the object is a `cudf.pandas` proxy `Index`.

> * `is_cudf_pandas_nd_array`: Detects if the object is a `cudf.pandas` proxy `ndarray`.

Author: ```suggestion
* `is_cudf_pandas_ndarray`: Detects if the object is a `cudf.pandas` proxy `ndarray`.
```

## How can I access the underlying GPU or CPU objects?

When working with `cudf.pandas` proxy objects, it is sometimes necessary to get true `cudf` or `pandas` objects that reside on GPU or CPU.
For example, this can be used to ensure that GPU-aware libraries that support both `cudf` and `pandas` can use the `cudf`-optimized code paths that keep data on GPU when processing `cudf.pandas` objects.
Otherwise, the library might use less-optimized CPU code because it thinks that the `cudf.pandas` object is a plain `pandas` dataframe.

The following methods can be used to retrieve the actual `cudf` or `pandas` objects:

- `as_gpu_object()`: This method returns the `cudf` object from the proxy.
- `as_cpu_object()`: This method returns the `pandas` object from the proxy.

> If `as_gpu_object()` is called on a proxy array, it will return a `cupy` array and `as_cpu_object` will return a `numpy` array.

Reviewer 2: What happens if it is called on something that is not a proxy type? If it is always the user's responsibility to check with the functions above, we should document that. Alternatively, we could do something like return `None` if it is not a proxy to support usage like:
```
if (ser := as_gpu_object(obj)) is not None:
    # Do unwrapped processing
```

but that feels unnecessarily complex.

Author: That's why we can use the `isinstance...` check we are introducing here no?

Reviewer 2: Yes I'm just wondering if there is a way to make this a one-step process, but I agree it ends up basically the same and doesn't improve much while adding complexity. No change required here.

Here is an example of how to use these methods:

```python
# Assuming `proxy_obj` is a cudf.pandas proxy object
cudf_obj = proxy_obj.as_gpu_object()
pandas_obj = proxy_obj.as_cpu_object()

# Now you can use `cudf_obj` and `pandas_obj` with libraries that are cudf or pandas aware
```

> Be aware that if `cudf.pandas` objects are converted to their underlying `cudf` or `pandas` types, the `cudf.pandas` proxy no longer controls them. This means that automatic conversion between GPU and CPU types and automatic fallback from GPU to CPU functionality will not occur.

Reviewer: ```suggestion
Be aware that if `cudf.pandas` objects are converted to their underlying `cudf` or `pandas` types, the `cudf.pandas` proxy no longer controls them.
This means that automatic conversion between GPU and CPU types and automatic fallback from GPU to CPU functionality will not occur.
```

blog](https://medium.com/rapids-ai/easy-cpu-gpu-arrays-and-dataframes-run-your-dask-code-where-youd-like-e349d92351d)) and the [RAPIDS Accelerator for Apache Spark](https://nvidia.github.io/spark-rapids/)
provides a similar configuration-based plugin for Spark.

## How do I know if an object is a `cudf.pandas` proxy object?

To determine if an object is a `cudf.pandas` proxy object, you can use the `isinstance_cudf_pandas` API. This function checks if the given object is a proxy object that wraps either a `cudf` or `pandas` object. Here is an example of how to use this API:

```python
from cudf.pandas import isinstance_cudf_pandas

obj = ...  # Your object here

> if isinstance_cudf_pandas(obj, pd.Series):

Reviewer 2: We should document if this is doing to work with `cudf.Series` if cudf is directly imported. It's fine if the answer is no, or if the answer is UB and we don't promise anything, we should just mention it either way.

- `as_gpu_object()`: This method returns the `cudf` object from the proxy.
- `as_cpu_object()`: This method returns the `pandas` object from the proxy.

> If, `as_gpu_object()` is called on a proxy array, it will return a `cupy` array and `as_cpu_object` will return a `numpy` array.

Reviewer: ```suggestion
If `as_gpu_object()` is called on a proxy array, it will return a `cupy` array and `as_cpu_object` will return a `numpy` array.
```
